src/training/data.py

```diff
@@ -30,8 +32,106 @@
 from open_clip.tokenizer import HFTokenizer
 
 
+def get_collate_fn(dataset, is_flava=False):
+    if is_flava:
+        assert isinstance(dataset.tokenize, HFTokenizer)
+        hf_tokenizer = dataset.tokenize.tokenizer
+        # The following special tokens are required for
+        # the data collator to work correctly.
+        assert getattr(hf_tokenizer, 'pad_token', False)
+        assert getattr(hf_tokenizer, 'mask_token', False)
+        mlm_collator = DataCollatorForLanguageModeling(
+            tokenizer=hf_tokenizer,
+            mlm=True,
+            mlm_probability=0.15,
+        )
+        pad_token_id = hf_tokenizer.pad_token_id
+
+    def collate(example_list):
+        image_list, neg_image_list, text_list = zip(*example_list)
+
+        image = torch.stack(image_list)
+        neg_image = torch.stack(neg_image_list)
+        text = torch.stack(text_list)
+        batch = {
+            'image': image,
+            'text': text,
+        }
+
+        if is_flava:
+            # Set up input mask to avoid attending to pad tokens
+            batch['text_input_mask'] = torch.where(text == pad_token_id, 0, 1)
+
+            # ITM
+            itm_prob = 0.1
+            itm_labels = torch.bernoulli(torch.ones(len(image)) * (1 - itm_prob))
+            corrupt_idx = torch.where(itm_labels == 0)
+            itm_image = image.clone()
+            itm_image[corrupt_idx] = neg_image[corrupt_idx]
+            batch.update({
+                'itm': {
+                    'label': itm_labels,
+                    'image': itm_image,
+                }
+            })
+
+            # MLM
+            assert mlm_collator is not None
+            mlm_input = mlm_collator(text_list)
+            batch.update({'mlm': mlm_input})
+
+        return batch
+
+    return collate
+
+
+class HFDataset(Dataset):
+    def __init__(
+        self,
+        name,
+        subset,
+        split,
+        transforms=None,
+        img_key=None,
+        caption_key=None,
+        tokenizer_name=None,
+        context_length=77,
```

Comment:
Is there a typo here with tokenizer_name? You are assigning tokenize below? Maybe you want tokenizer instead everywhere?